### PR TITLE
verify modules before compiling or publishing

### LINKED ISSUE #26: verification of module references in builds
It's possible to create links in builds to modules which do not resolve. This is fine during setup of the build (for example, you can specify that the compiler to use should be `scala/compiler` before you've imported a workspace containing the Scala compiler. But we need to check that everything resolves before attempting to run a build or publish a workspace.

References to modules which do not resolve should be marked with a bright red `!!` (or similar) until they are fixed.

### PR DESCRIPTION
**Pull request by mzarnowski**
closes #26

### CODE REVIEW

File: src/build/build.scala [hunk 1]
```diff
@@ -390,6 +392,9 @@ object LayerCli {
       cli           <- cli.hint(TagArg, suggestedTags)
       invoc         <- cli.read()
       io            <- invoc.io()
+      hierarchy     <- schema.hierarchy(io, layout.pwd, layout)
+      universe      <- hierarchy.universe
+      _             <- Verifier.verifyUniverse(universe)
```
Comment (propensive) [inline src/build/build.scala]: This isn't quite necessary. We only need to verify all the modules in all the projects in all the schemas in the current layer. We don't need to verify the entire universe. But we do need the universe to check that the references resolve.

File: src/core/fury/module/Verifier.scala [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import fury.error._
+import fury._
+
+import scala.util.{Failure, Success}
+
+object Verifier {
+
+  def verifyUniverse(universe: Universe): Outcome[Set[ModuleId]] = {
+    val definedModules = for {
+      schema  <- universe.schemas.values
```
Comment (propensive) [inline src/core/fury/module/Verifier.scala]: So this should be just `layer.schemas`.

File: src/core/fury/module/Verifier.scala [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      project <- schema.projects
+      module  <- project.modules
+    } yield module
+
+    definedModules.foldLeft(~Set[ModuleId]()) {
```
Comment (propensive) [inline src/core/fury/module/Verifier.scala]: This fold is fine, but a recursive definition might be easier to work with. No need to change it.

File: src/core/fury/module/Verifier.scala [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      case (Success(verifiedModules), module) =>
+        verifyModule(universe, module, verifiedModules)
+      case (f, _) => f
+    }
+  }
+
+  def verifyModule(universe: Universe, module: Module): Outcome[Set[ModuleId]] =
```
Comment (propensive) [inline src/core/fury/module/Verifier.scala]: I think this is more complicated than it needs to be. Don't worry - that's a good thing. Here's some pseudocode for what we need to do
```
for {
  schema <- layer.schemas
  project <- schema.projects
  module <- project.modules
  dep <- module.dependencies + module.compiler
 } yield universe.projects(dep.projectId).contains(dep.moduleId)
```

Then make this into a fold or a recursive method.

**Review by propensive — CHANGES_REQUESTED** [review]

**Comment by odisseus** [comment]
This pull request has become unmergeable.

**Pull request closed without merging**